File: ParadigmaComp/15Questoes/python/questao7.py
# Associação
class Professor:
    def __init__(self, nome):
        self.nome = nome
        self.escolas = []

    def adicionar_escola(self, escola):
        if escola not in self.escolas:
            self.escolas.append(escola)

    def exibir_escolas(self):
        return [escola.nome for escola in self.escolas]

class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []

    def adicionar_professor(self, professor):
        if professor not in self.professores:
            self.professores.append(professor)
            professor.adicionar_escola(self)

    def exibir_professores(self):
        return [professor.nome for professor in self.professores]

# Exemplo de uso
professor1 = Professor("Dr. João")
professor2 = Professor("Prof. Maria")

escola1 = Escola("Escola A")
escola2 = Escola("Escola B")

escola1.adicionar_professor(professor1)
escola1.adicionar_professor(professor2)
escola2.adicionar_professor(professor1)

print(escola1.exibir_professores())  # Professores da Escola A
print(escola2.exibir_professores())  # Professores da Escola B
print(professor1.exibir_escolas())   # Escolas onde o Dr. João leciona

File: ParadigmaComp/15Questoes/golang/questao7.go
package main

// Associação

import "fmt"

// Estrutura Professor
type Professor struct {
	Nome    string
	Escolas []*Escola
}

func (p *Professor) AdicionarEscola(escola *Escola) {
	p.Escolas = append(p.Escolas, escola)
}

// Estrutura Escola
type Escola struct {
	Nome       string
	Professores []*Professor
}

func (e *Escola) AdicionarProfessor(professor *Professor) {
	e.Professores = append(e.Professores, professor)
	professor.AdicionarEscola(e)
}

func main() {
	// Exemplo de uso
	professor1 := &Professor{Nome: "Dr. João"}
	professor2 := &Professor{Nome: "Prof. Maria"}

	escola1 := &Escola{Nome: "Escola A"}
	escola2 := &Escola{Nome: "Escola B"}

	escola1.AdicionarProfessor(professor1)
	escola1.AdicionarProfessor(professor2)
	escola2.AdicionarProfessor(professor1)

	// Professores da Escola A
	for _, professor := range escola1.Professores {
		fmt.Println(professor.Nome)
	}

	// Professores da Escola B
	for _, professor := range escola2.Professores {
		fmt.Println(professor.Nome)
	}

	// Escolas onde o Dr. João leciona
	for _, escola := range professor1.Escolas {
		fmt.Println(escola.Nome)
	}
}
